[yup] add generic to .test()

ejose19 opened the pull request:
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [X] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

iansan5653 commented:
This could lead to some hidden errors. For example:
```ts
const schema = yup.string().defined().test(x => x === x.toUpperCase()); // x has type string so TS says this is OK
console.log(schema.isValidSync("abc")); // false - this is good. isValidSync returns boolean, not error
console.log(schema.isValidSync(undefined)); // errors! should return false but undefined.toUpperCase is not a function
```

Now, the current implementation also allows for hidden errors, but it makes you be more explicit about them:
```ts
const schema = yup.string().defined().test(x => x === x.toUpperCase()); // x has type any to TS says this is OK, but warns for implicit any
const schema = yup.string().defined().test((x: String) => x === x.toUpperCase()); // TS says this is OK but at least you know you asserted this
```

Instead what I would support is changing the `any` that is currently there to `unknown`:
```ts
const schema = yup.string().defined().test(x => x === x.toUpperCase()); // not allowed if x has type unknown
```
Instead, you'd have to do:
```ts
const schema = yup.string().defined().test(x => typeof x === "string" && x === x.toUpperCase()); // this is best
```

ejose19 commented:
@iansan5653 Why is .test() running if there's a previous .defined() that should have failed? When I wrote this I thought they were run in definition order so that error you mention shouldn't happen.

iansan5653 commented:
That's what I would have thought too, but if you think about it Yup has to be able to run everything in order. For example:

```ts
yup.string().defined().ensure().isValidSync(undefined); // true because result is ""
```

If checking stopped at the failed `defined`, then it wouldn't hit the `ensure` (which has the effect of transforming `undefined -> ""`) and it would see it as invalid.

ejose19 commented:
Yes, guess it's a limitation until a sequential validation is implemented.

I've checked the type definitions for .defined(), but from what you mention then we've a tricky one, both .required() and .defined() removes undefined from the value, but that's incorrect as test are not run sequentially, and if you remove that from type definitions then typescript would have alerted you that x "is possibly undefined".

ejose19 commented:
I've updated the PR, I think this new proposal suits it better.

Instead of removing the narrowing on .required() / .defined() / .nullable() (which would break integration of schema with user interfaces/types) I added (undefined | null) to the generic on .test() so typescript properly gives the warning that the value may be undefined or null, and if [docs](https://github.com/jquense/yup#mixedtestname-string-message-string--function-test-function-schema) are correct on "Tests are run after any object is cast" then only (T | null | undefined) are the only possible values at runtime.

iansan5653 commented on types/yup/index.d.ts:
```diff
@@ -361,15 +361,15 @@ export interface ObjectSchema<T extends object | null | undefined = object | und
 }
 
-export type TestFunction = (
+export type TestFunction<T = any> = (
```
It would be safer for `MixedSchema` to make this default to `unknown`

iansan5653 commented on types/yup/index.d.ts:
```diff
@@ -147,8 +147,8 @@ export interface StringSchema<T extends string | null | undefined = string | und
         arrayOfValues: ReadonlyArray<U | Ref>,
         message?: MixedLocale['oneOf'],
     ): StringSchema<MaintainOptionality<T, U>>;
+    test(name: string, message: TestOptionsMessage, test: TestFunction<T | undefined | null>): this;
```
I would add a comment here explaining that `| undefined | null` is added intentionally because the test will still run for undefined/null input whether or not they're set to be allowed. Otherwise inevitably someone else will submit a PR to remove that.

ejose19 commented:
@iansan5653 I've updated the PR with the changes requested. Thanks for checking it out.

ejose19 commented:
Ready to merge

HansAarneLiblik commented:
Hey!

Shouldn't 
`test(options: TestOptions<Record<string, any>>): this;` 
be
`test(options: TestOptions<T | undefined | null>): this;` 
for `BasicArraySchema` (probably something similar to other Schemas as well)
?

Also I think you missed generic `<P>` for `test: TestFunction;` in `interface BaseTestOptions<P extends Record<string, any>>`

This is breaking our build currently

ejose19 commented:
@HansAarneLiblik TestOptions receive an object in the form of

```ts
Options = {
  // unique name identifying the test
  name: string;
  // test function, determines schema validity
  test: (value: any) => boolean;
  // the validation error message
  message: string;
  // values passed to message for interpolation
  params: ?object;
  // mark the test as exclusive, meaning only one of the same can be active at once
  exclusive: boolean = false;
}
```
so you can't replace it with T | null | undefined (same with generic P, it's referring to options type not test value type). You could do this however if you implement 2 generics, <T, P> where P remain as it's currently and T is the type of the property.

Regarding the break of your build, I'm guessing it's because test value type now defaults to unknown, as @iansan5653 suggested to make tests where generic wasn't implemented/reachable safer.
